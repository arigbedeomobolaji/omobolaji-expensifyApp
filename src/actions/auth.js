import {firebase, googleAuthProvider }  from "../firebase/firebase"

export const login = (uid) => ({
 type: "LOGIN",
  uid
})

// export const asyncLogin = () => {
//  return (dispatch) => {
//   return firebase.auth().onAuthStateChanged((user) => {
//    if (user) {
//     dispatch(login(user.uid))
//    }
//   })
//  } 
// }

export const startLogin = () => {
 return () => {
  return firebase.auth().signInWithPopup(googleAuthProvider)
 }
}

export const logout = () => ({
 type: "LOGOUT"
})

export const startLogout = () => {
 return () => {
  return firebase.auth().signOut()
 }
}