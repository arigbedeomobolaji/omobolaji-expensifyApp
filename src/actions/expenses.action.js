import { v4 as uuid } from "uuid"
import database from "../firebase/firebase"

// What's happening in our code before async redux
// 1. Component calls action generator
// 2. action generation returns an object
// 3. component dispatches object
// 4. redux store changes

// What happens when we use async redux
// 1. Component calls action generator
// 2. action generator returns a function
// 3. component dispatches function (?)
// 4. function run (has the ability to dispatch other actions and do whatever it wants)

//ADD_EXPENSE
export const addExpense = (expense = {}) => ({
 type: "ADD_EXPENSE",
 expense
})

// async addExpense
export const asyncAddExpense = (expenseData = {}) => {
  // this is the function returned by the redux action generator
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData
    const expense = { description, note, amount, createdAt }
    return database.ref(`users/${uid}/expenses`).push(expense).then((snapshot) => {
      // What I'm doing here is that I'm updating the redux store so that as the database changes reduc store also change
      dispatch(addExpense({
        id: snapshot.key,
        ...expense
      }))
    })
  }
}

//REMOVE_EXPENSE
export const removeExpense = (id) => ({
 type: "REMOVE_EXPENSE",
 id
})

export const asyncRemoveExpense = (id) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense(id))
    })
  }
}

///EDIT_EXPENSE
export const editExpense = (id, updates) => ({
 type: "EDIT_EXPENSE",
 id,
 updates
})

export const asyncEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates))
    })
  }
}

// SET_EXPENSES => fetches expenses in the databse and save it as an array
export const setExpenses = (expenses) => ({
  type: "SET_EXPENSES",
  expenses
})

export const asyncSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses`).once("value").then((snapshot) => {
      const expenses = []
      // const firebaseExpenses = snapshot.val()
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
      dispatch(setExpenses(expenses))
    })
  }
};