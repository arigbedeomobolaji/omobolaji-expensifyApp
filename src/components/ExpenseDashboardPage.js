//jshint ignore:start

import React from "react"
import ConnectExpenseList from "./ExpenseList"
import ConnectedExpenseListFilter from "./ExpenseListFilter"
import ConnectedExpenseSummary from "./ExpenseSummary"

const ExpenseDashboardPage = () => (
 <div>
  <ConnectedExpenseSummary />
  <ConnectedExpenseListFilter />
  <ConnectExpenseList />
 </div>
)

export default ExpenseDashboardPage