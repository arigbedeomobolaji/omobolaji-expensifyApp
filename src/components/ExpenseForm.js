//jshint ignore:start
import React from "react"
import moment from "moment"
import 'react-dates/initialize'
import { SingleDatePicker } from "react-dates"


export default class ExpenseForm extends React.Component {
 constructor(props){
  super(props)
  this.state = {
   description: props.expense ? props.expense.description : "",
   note: props.expense ? props.expense.note :"",
   amount: props.expense ? (props.expense.amount/100).toString() : "",
   createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
   isCalendarFocused: false,
   error: ""
  }
 }

 onDescriptionChange = (e) => {
  const description = e.target.value
  this.setState(() => ({
   description
  }))
 }

 onNoteChange = (e) => {
  const note = e.target.value
  this.setState(() => ({
   note
  }))
 }

 onAmountChange = (e) => {
  const amount = e.target.value
  if (amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
   this.setState(() => ({
    amount: amount
   }))
  }
 }

 onDateChange = (createdAt) => {
  if (createdAt) {
   this.setState(() => ({ createdAt }))
  }
 }
 

 onCalendarFocusedChange = ({ focused }) => {
  this.setState(() => ({
   isCalendarFocused: focused
  }))
 }

 onSubmit = (e) => {
  e.preventDefault()
  
  if (!this.state.description || !this.state.amount) {
   // Render Error 
   this.setState(() => ({
    error: "Please provide description and amount."
   }))
  } else {
   // clear error
   this.setState(() => ({
    error: ""
   }))
   // We created the expense props simply because we want to reuse this form on AddExpense and EditExpense
   // We can also call a function as a props
   this.props.onSubmit({
    description: this.state.description,
    amount: parseFloat(this.state.amount, 10) * 100,
    createdAt: this.state.createdAt.valueOf(),
    note: this.state.note
   })
  }
 }

 render() {
  return (
    <form className="form" onSubmit={this.onSubmit}>
     {!!this.state.error && <p className="form__error">{this.state.error}</p>}
     <input
      type="text"
      className="text-input"
      placeholder="Description"
      autoFocus
      value={this.state.description}
      onChange={this.onDescriptionChange}
     />
     <input 
      type="text"
      className="text-input"
      placeholder="Amount"
      value={this.state.amount}
      onChange={this.onAmountChange}
     />
     <SingleDatePicker 
      date={this.state.createdAt}
      onDateChange={this.onDateChange}
      focused={this.state.isCalendarFocused}
      onFocusChange={this.onCalendarFocusedChange}
      numberOfMonths={1}
      isOutsideRange={() => false}
     />
     <textarea
      className="textarea"
      placeholder="Add a Note (Optional)"
      value={this.state.note}
      onChange={this.onNoteChange}
     >
     </textarea>
     <div>
       <button className="button button--secondary" type="submit"> { !!this.props.buttonName ? this.props.buttonName : "Add Expense"}</button>
     </div>
    </form>
  )
 }
}