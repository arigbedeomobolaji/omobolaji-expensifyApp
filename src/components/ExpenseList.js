//jshint ignore:start

import React from "react"
import { connect } from "react-redux"
import ExpenseListItem from "./ExpenseListItem"
import selectExpenses from "../selectors/expenses.selector"

export const ExpenseList = (props) => (
 <div>
  <h1>ExpenseList</h1>
  {
   props.expenses.length === 0 ? ( <p>No Expense to show</p>) : (
    
  <table>
   <thead>
    <tr>
     <th>Description</th>
     <th>Amount</th>
     <th>createdAt</th>
    </tr>
   </thead>
   <tbody>
    {props.expenses.map(
     (expense) => {
      return <ExpenseListItem key={expense.id} {...expense} />
     }
    )}
   </tbody>
  </table>
   )
  }
 </div>
)

const mapStateToProps = (state) => ({
 expenses: selectExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpenseList)