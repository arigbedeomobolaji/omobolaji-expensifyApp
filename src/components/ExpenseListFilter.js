//jshint ignore: start
import React from "react"
import { connect } from "react-redux"
import "react-dates/initialize"
import { DateRangePicker } from "react-dates"
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from "../actions/filters.action"
import "react-dates/lib/css/_datepicker.css"

export class ExpenseListFilter extends React.Component {

  state = {
  calendarFocused: null
 }

 onDatesChange = ({ startDate, endDate }) => {
  this.props.setStartDate(startDate)
  this.props.setEndDate(endDate)
  }
  
  onFocusChange = (calendarFocused) => this.setState(() => (
    { calendarFocused }
  ))
    
  onTextFilterChange = (e) => {
    this.props.setTextFilter(e.target.value)
  }
    
  onSortByChange = (e) => {
    e.target.value === "date"? this.props.sortByDate() : this.props.sortByAmount()
  }

  render() {  
    return (
    <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
            type="text"
            className="text-input"
            placeholder="Search expenses"
            //Any time you see value like below they are called CONTROLLED INPUT ==> Input controlled by Javascript
            value={this.props.filters.text}
            onChange={this.onTextFilterChange} 
            />
          </div>
          <div className="input-group__item">
            <select
            className="select"
            onChange={this.onSortByChange}
            >
            <option value="date">Date</option>
            <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker 
            startDate={this.props.filters.startDate}
            startDateId={"myStartDateId"}
            endDate={this.props.filters.endDate}
            endDateId={"myEndDateId"}
            onDatesChange={this.onDatesChange}
            focusedInput={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
            showClearDates={true}
            />
          </div>
      </div>
    </div>
    )
 }
}

const mapStateToProps = (state) => {
 return {
  filters: state.filters
 }
}

const mapDispatchToProps = (dispatch) => ({
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter)