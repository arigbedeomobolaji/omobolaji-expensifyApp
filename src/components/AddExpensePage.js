//jshint ignore:start
import React from "react"
import {connect} from "react-redux"
import ExpenseForm from "./ExpenseForm"
import {asyncAddExpense} from "../actions/expenses.action.js"

// History is passed into AddExpensePage becus AddExpensagePage is registered to a Route
export class AddExpensePage extends React.Component{
 onSubmit = (expense) => {
  this.props.asyncAddExpense(expense)
  this.props.history.push("/dashboard")
 }
 render() {
  return (
   <div>
    <div className="page-header">
     <div className="content-container">
      <h1 className="page-header__title">Add Expense</h1>
     </div>
    </div>
    <div className="content-container">
     <ExpenseForm
      onSubmit={this.onSubmit}
     />
    </div>
   </div>
  )
 }
}

const mapDispatchToProps = (dispatch) => ({
 asyncAddExpense: (expense) => dispatch(asyncAddExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)

// mapDispatchToProps is a way to return your dispatcher functions allowing you to abstract them away from the component itself.
// to avoid inline functions we are going to switch our stateless function to a class based function