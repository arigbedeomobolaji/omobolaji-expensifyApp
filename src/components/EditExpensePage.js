//jshint ignore:start
import React from "react"
import { connect } from "react-redux"
import ExpenseForm from "./ExpenseForm"
import {asyncEditExpense, asyncRemoveExpense} from "../actions/expenses.action"

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    // props.dispatch(editExpense(props.expense.id, expense))
    this.props.asyncEditExpense(this.props.expense.id, expense)
    this.props.history.push("/dashboard")
  }

  handleRemoveExpense = (e) => {
    this.props.asyncRemoveExpense(this.props.expense.id)
    this.props.history.push("/dashboard")
  }

  render() {
    return (
      <div>
        <h1>Expensify App | Edit</h1>
      <ExpenseForm
        expense={this.props.expense}
        onSubmit={this.onSubmit}
      />
      <button onClick={this.handleRemoveExpense}
      >Remove</button>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)   
})

const mapDispatchToProps = (dispatch) => ({
  asyncEditExpense: (id, expense) => {
    dispatch(asyncEditExpense(id, expense))
  },
  asyncRemoveExpense: (id) => {
    dispatch(asyncRemoveExpense(id))
  }
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
