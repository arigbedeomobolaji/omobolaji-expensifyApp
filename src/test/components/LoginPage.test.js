//jshint ignore:start
import React from "react"
import { shallow } from "enzyme"
import { LoginPage } from "../../components/LoginPage"
import { startLogout } from "../../actions/auth"

test("Should render LoginPage correctly", () => {
 const wrapper = shallow(<LoginPage startLogin={jest.fn()} />)
 expect(wrapper).toMatchSnapshot()
})

test("should call startLogin on click", () => {
 const startLogin = jest.fn()
 const wrapper = shallow(<LoginPage startLogin={startLogin} />)
 wrapper.find("button").simulate("click")
 expect(startLogin).toHaveBeenLastCalledWith()
})

