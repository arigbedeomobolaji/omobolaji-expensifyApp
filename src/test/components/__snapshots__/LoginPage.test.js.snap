// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render LoginPage correctly 1`] = `
<div>
  <button
    onClick={[MockFunction]}
  >
    Login
  </button>
</div>
`;
